src/Distribution/Nixpkgs/Meta.hs

```diff
@@ -85,6 +92,54 @@ nixpkgsPlatformFromCabal (Platform arch os) = "\"" ++ nixArch ++ "-" ++ nixOs ++
             OSX -> "darwin" -- rendered as osx by default
             _   -> CabalPretty.prettyShow os
 
+-- | Obtain a 'NixpkgsPlatform' from a string representation intended for config
+--   files.
+--
+--   * Every string starting with @lib.platforms.@ or @platforms.@ is
+--     parsed into 'NixpkgsPlatformGroup'.
+--   * All other strings are attempted to be interpreted as a system tuple of
+--     the form @ARCH-KERNEL@ and parsed into 'NixpkgsPlatformSingle'.
+--
+--   If none of these formats match the input 'String', 'Nothing' is returned.
+--   A 'Just' result thus only indicates that the format of the platform is
+--   sound — 'nixpkgsPlatformFromString' does /not/ check if the parsed platform
+--   actually exists.
+--
+--   __Note__: 'nixpkgsPlatformFromString' is /not/ the inverse operation for
+--   'NixpkgsPlatform'\'s 'Pretty' instance. It is not intended for parsing Nix
+--   expressions.
+--
+--   >>> nixpkgsPlatformFromString "x86_64-netbsd"
+--   Just (NixpkgsPlatformSingle (Platform X86_64 NetBSD))
+--   >>> nixpkgsPlatformFromString "platforms.riscv"
+--   Just (NixpkgsPlatformGroup (Identifier "riscv"))
+--   >>> nixpkgsPlatformFromString "garbage"
+--   Nothing
+nixpkgsPlatformFromString :: String -> Maybe NixpkgsPlatform
+nixpkgsPlatformFromString s = platformGroup <|> singlePlatform
+  where platformGroup = do
+          -- also accept "platform." as prefix to save some typing
+          name <- stripPrefix "lib.platforms." s <|> stripPrefix "platforms." s
+          Just $ NixpkgsPlatformGroup (ident # name)
+
+        -- we accept anything as a platform tuple which is two non-empty
+        -- strings separated by a '-'
+        splitPlatformTuple = filter (not . null) . splitOn "-"
+        singlePlatform =
+          case splitPlatformTuple s of
+            [arch, os] -> Just $ NixpkgsPlatformSingle
```

Comment:
Turns out this is not quite correct: Platform tuples (or doubles as they are called in nixpkgs' lib) can have three components separated by dashes: E. g. `riscv64-none-elf` should be parsed as `(OtherArch "riscv64", OtherOS "none-elf")`.

Thus we should use `Data.List.break` instead of `splitOn`. One problematic consequence of this is that it gets much harder to sanity-check the tuple since basically anything with more than one dash is technically valid. Even worse: There's no way to distinguish tuples and triples without keeping a massive lookup table to validate the individual parts!

This is not relevant for the (originally) intended domain of this code, the `hackage2nix` config. However, ideally we'd reuse this code for `cabal2nix`'s command line parsing where this is sadly relevant. I'm not sure which direction to go with this yet, just writing this down as an update.

Reply:
Now works correctly. The problem in `cabal2nix`'s command line interface is avoided for now by continuing to use the old code. We are always free to be stricter there, so I don't think we need to concern us with this problem here.